File: src/languages/hebrew/lexicon.ts
/**
 * Hebrew (he) pronunciation lexicon — a LOOKUP layer for the unvocalized (neural) path. A known skeleton maps to a
 * stored niqqud (vocalized) reading, which we render through the rule g2p (`phonemizeWord`) — so the lexicon
 * output is always in OUR canonical convention and auto-tracks g2p rule changes (the glottal rule etc.), never
 * importing an external IPA convention. Homographs and OOV words are NOT in the lexicon and fall through to the
 * context-aware tagger. See he-lexicon.tsv + tools/hebrew/build-lexicon.py.
 */

import { phonemizeWord } from "./hebrew.ts";
import { dataFile } from "../../core/dataPath.ts";
import { readDataText } from "../../core/dataSource.ts";

let cache: Map<string, string> | undefined; // skeleton → niqqud

function load(): Map<string, string> {
    if (cache) return cache;
    cache = new Map();
    try {
        const key = dataFile(import.meta.url, "he-lexicon.tsv");
        for (const line of readDataText(key).split("\n")) {
            if (!line || line.startsWith("#")) continue;
            const [skel, niqqud] = line.split("\t");
            if (skel && niqqud) cache.set(skel, niqqud);
        }
    } catch { /* no lexicon shipped → empty (every lookup misses, tagger handles all) */ }
    return cache;
}

/** The lexicon reading of an unvocalized word in OUR convention, or undefined if not a known non-homograph word. */
export function lexiconLookup(skeleton: string): string | undefined {
    const niqqud = load().get(skeleton);
    return niqqud === undefined ? undefined : phonemizeWord(niqqud);
}

File: csharp/Vernacula.Phonemizer/Languages/Hebrew/Lexicon.cs
/**
 * Hebrew (he) pronunciation lexicon — a LOOKUP layer for the unvocalized (neural) path: a known skeleton
 * maps to a stored niqqud reading, rendered through the rule g2p so it stays in OUR convention.
 * Ported from src/languages/hebrew/lexicon.ts — see that file for the provenance.
 */
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Hebrew;

public static class Lexicon
{
    private static IReadOnlyDictionary<string, string>? cache; // skeleton → niqqud

    private static IReadOnlyDictionary<string, string> Load() =>
        // `optional: true` is the TS's `catch { }` — no lexicon shipped → empty, every lookup misses.
        cache ??= LoadTsv.LoadTsvMap("languages/hebrew", "he-lexicon.tsv", optional: true);

    /** The lexicon reading of an unvocalized word in OUR convention, or null if not a known non-homograph. */
    public static string? LexiconLookup(string skeleton) =>
        Load().TryGetValue(skeleton, out var niqqud) ? HebrewPhonemizer.PhonemizeWord(niqqud) : null;
}
